Added report generation based in data tiers and campaigns

andrius-k opened the pull request.

vkuznet commented on src/bash/report_campaigns/aggregate_campaign_tier:
```diff
@@ -0,0 +1,45 @@
+# Remove previous data first
+hadoop fs -rm -r $location
+
+PYTHONPATH=$(pwd)/../../python ../../../bin/run_spark /reports/aggregate_campaign_tier.py --fout=$hdir --yarn --verbose --date=$date
```
You should fix all relative paths and never use them. They implies certain directory layout and will easily break when this is changed which will make hard to maintain the scripts. Instead you should find out how to provide full path to script, e.g. see example here:
https://github.com/vkuznet/CMSSpark/blob/master/bin/cron4phedex#L40-L47

vkuznet commented on src/bash/report_campaigns/aggregate_campaign_tier:
```diff
@@ -0,0 +1,45 @@
+echo 'Aggregating for date: '$date
+
+# Results will be put here
+location=/cms/users/$USER/campaign_tier
```
it is better to provide output area as external parameter to the script, otherwise it would be impossible to run this script multiple times since it will yield results in the same area. For instance, I may need to run this scripts to check the output and therefore want to put results into different output areas or I want to preserve my output area, etc.

vkuznet commented on src/bash/report_campaigns/aggregate_campaigns:
```diff
@@ -0,0 +1,59 @@
+echo 'Aggregating for date: '$date
+
+# Results will be put here
+location=/cms/users/$USER/campaigns
```
same argument as in aggregate_campaign_tier

vkuznet commented on src/bash/report_campaigns/aggregate_campaigns:
```diff
@@ -0,0 +1,59 @@
+# Remove previous data first
+hadoop fs -rm -r $location
+
+PYTHONPATH=$(pwd)/../../python ../../../bin/run_spark /reports/aggregate_campaigns.py --fout=$hdir --yarn --verbose --date=$date
```
same argument as in aggregate_campaign_tier

vkuznet commented on src/bash/report_campaigns/aggregate_campaigns:
```diff
@@ -0,0 +1,59 @@
+#!/bin/sh
+
+# Parse date argument
+date=20170228
```
should be input parameter, no hardcoded values.

vkuznet commented on src/bash/report_campaigns/aggregate_campaign_tier:
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
+
+# Parse date argument
+date=20170228
```
should be input parameters, no hardcoded values should be allowed

vkuznet commented on src/bash/report_tiers/aggregate_dbs:
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
+
+# Parse date argument
+date=20170228
```
see aggregate_campaign_tier

vkuznet commented on src/bash/report_tiers/aggregate_dbs:
```diff
@@ -0,0 +1,45 @@
+echo 'Aggregating for date: '$date
+
+# Results will be put here
+location=/cms/users/$USER/dbs_datasets
```
see aggregate_campaign_tier

vkuznet commented on src/bash/report_tiers/aggregate_dbs:
```diff
@@ -0,0 +1,45 @@
+# Remove previous data first
+hadoop fs -rm -r $location
+
+PYTHONPATH=$(pwd)/../../python ../../../bin/run_spark /reports/aggregate_dbs.py --fout=$hdir --yarn --verbose
```
see aggregate_campaign_tier

vkuznet commented on src/bash/report_tiers/aggregate_phedex:
```diff
@@ -0,0 +1,39 @@
+#!/bin/sh
+
+# Parse date argument
+date=20170228
```
see aggregate_campaign_tier

vkuznet commented on src/bash/report_tiers/aggregate_phedex:
```diff
@@ -0,0 +1,39 @@
+# Remove previous data first
+hadoop fs -rm -r $location
+
+PYTHONPATH=$(pwd)/../../python ../../../bin/run_spark /reports/aggregate_phedex.py --fout=$hdir --yarn --verbose --date=20170228
```
see aggregate_campaign_tier

vkuznet commented on src/bash/report_tiers/aggregate_phedex:
```diff
@@ -0,0 +1,39 @@
+
+# concatenate all parts except header
+header=`cat phedex_df.csv`
+cat phedex_datasets/2017/02/28/part* | grep -v $header >> phedex_df.csv
```
no hardcoded values should be allowed

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaign_tier.py:
```diff
@@ -0,0 +1,200 @@
+from CMSSpark.spark_utils import spark_context, split_dataset
+from CMSSpark.utils import elapsed_time
+
+LIMIT = 100
```
make it as an option instead of hardcoded value

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaigns.py:
```diff
@@ -0,0 +1,264 @@
+from CMSSpark.spark_utils import spark_context, split_dataset
+from CMSSpark.utils import elapsed_time
+
+LIMIT = 100
```
make it as input parameter, no hardcoded values

vkuznet commented on src/python/CMSSpark/reports/visualize_campaigns.py:
```diff
@@ -0,0 +1,213 @@
+report = ''
+
+def append_report(lines):
+    global report
```
remove **all** globals and never use them, they always lead to hard maintenance in a code base

vkuznet commented on src/python/CMSSpark/reports/visualize_tiers.py:
```diff
@@ -0,0 +1,215 @@
+report = ''
+
+def append_report(lines):
+    global report
```
remove globals

vkuznet commented on src/python/CMSSpark/reports/visualize_tiers.py:
```diff
@@ -0,0 +1,215 @@
+import operator
+import argparse
+
+PHEDEX_PLOTS_PATH = 'phedex_plots/'
```
should be part of options with these default values.

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaigns.py:
```diff
@@ -137,22 +137,23 @@ def run(fout, date, yarn=None, verbose=None, patterns=None, antipatterns=None, i
-                   .withColumnRenamed('f_file_size', 'size')\
+    # d_dataset_id, d_dataset, dataset_access_type
+    dbs_df = ddf_df.join(daf_df, ddf_df.d_dataset_access_type_id == daf_df.dataset_access_type_id)\
+                   .drop(ddf_df.d_dataset_access_type_id)\
```
Could you please check that .drop here does not do any harm. My worry is that you just joined two tables and immediately drop the foreign key you're joining. I would suggest dropping columns at the end when you construct all joins.

andrius-k replied:
I just checked and it appears that it makes no difference. I tried dropping columns in the new expression after join is completed and the number of campaigns is the same.

vkuznet replied:
ok, I thought it was worth of checking anyhow. Let me know when you'll be done with the rest of comments.

andrius-k replied:
Ok, I'm now working on few more attempts to solve the leftover campaigns issue. I'll report my findings shortly.

vkuznet commented:
Andrius, let me know when you're ready with this that I can review again from top to bottom. And, based on today's meeting outcome please prepare README_leftover.md (or whatever you'll call it) markdown with instructions how to run your leftover script.

andrius-k commented:
Dear Valentin,

ok, I will continue working on fixing these issues now. I should be finished very soon as I have completed most of them already. 

And, sure, I will create a leftovers readme as we discussed in the meeting. By the way, do you want me to create another report for leftovers as well? Because now I only run the script that generates two data frames. I think we should come up with another visualization technique because in order to commit report, script runner (user) needs to authenticate with my github account. 

Regards, 

Andrius

> On 20 Mar 2018, at 20:29, Valentin Kuznetsov <notifications@github.com> wrote:
> 
> Andrius, let me know when you're ready with this that I can review again from top to bottom. And, based on today's meeting outcome please prepare README_leftover.md (or whatever you'll call it) markdown with instructions how to run your leftover script.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

vkuznet commented:
Andrius,
please create report for leftovers as well. As I mentioned we'll need to move
our data to CERN MONIT for visualization. Hopefully we'll discuss this next
week.
Best,
Valentin.

On  0, Andrius <notifications@github.com> wrote:
> Dear Valentin,
> 
> ok, I will continue working on fixing these issues now. I should be finished very soon as I have completed most of them already. 
> 
> And, sure, I will create a leftovers readme as we discussed in the meeting. By the way, do you want me to create another report for leftovers as well? Because now I only run the script that generates two data frames. I think we should come up with another visualization technique because in order to commit report, script runner (user) needs to authenticate with my github account. 
> 
> Regards, 
> 
> Andrius
> 
> > On 20 Mar 2018, at 20:29, Valentin Kuznetsov <notifications@github.com> wrote:
> > 
> > Andrius, let me know when you're ready with this that I can review again from top to bottom. And, based on today's meeting outcome please prepare README_leftover.md (or whatever you'll call it) markdown with instructions how to run your leftover script.
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> > 
> 
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/vkuznet/CMSSpark/pull/12#issuecomment-374735355

andrius-k commented:
Dear Valentin,

I have short question: should I consider all DBS instances (GLOBAL, PHYS01, PHYS02, PHYS03) when doing leftover aggregation?

Regards,

Andrius

vkuznet commented:
Make it as external parameter such that user can choose either specific instance or 'all'.

On  0, Andrius <notifications@github.com> wrote:
> Dear Valentin,
> 
> I have short question: should I consider all DBS instances (GLOBAL, PHYS01, PHYS02, PHYS03) when doing leftover aggregation?
> 
> Regards,
> 
> Andrius
> 
> 
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/vkuznet/CMSSpark/pull/12#issuecomment-374976461

vkuznet commented:
Andrius, could you please rebase your pull request with new master. I re-factored code and included new conf module which defines OptionParser class. Almost all code had duplication of this class and I made necessary changes to handle this better. Basically now you need to import OptionParser and adjust it to your setup of options. See how it is done in 
https://github.com/vkuznet/CMSSpark/blob/master/src/python/CMSSpark/dbs_phedex.py#L22-L23
https://github.com/vkuznet/CMSSpark/blob/master/src/python/CMSSpark/dbs_phedex.py#L119-139

Since your PR is not yet merged I let you to adjust your code base as well to use CMSSpark.conf module. Please make necessary changes.

Thanks,
Valentin.

vkuznet commented:
because of my changes you'll also need to resolve conflicts in your utils.py (sorry about that)

andrius-k commented:
Valentin, I corrected all issues pointed by you and I added leftovers report generation script and instructions on how to run it. If there are any more issues or anything else you want me to change - please let me know.

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaign_tier.py:
```diff
@@ -0,0 +1,193 @@
+# CMSSpark modules
+from CMSSpark.spark_utils import dbs_tables, phedex_tables, print_rows
+from CMSSpark.spark_utils import spark_context, split_dataset
+from CMSSpark.utils import elapsed_time
```
replace this with
from CMSSpark.utils import info
from CMSSpark.conf import OptionParser

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaign_tier.py:
```diff
@@ -0,0 +1,193 @@
+
+CAMPAIGN_TIER_TIME_DATA_FILE = 'spark_exec_time_campaign_tier.txt'
+
+def get_options():
```
adjust this function to use OptionParser since it provides almost all generic args we need to run a job

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaign_tier.py:
```diff
@@ -0,0 +1,193 @@
+
+    ctx.stop()
+
+def main():
```
Add `@info` decorator for main since it keeps track of elapsed time now.

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaign_tier.py:
```diff
@@ -0,0 +1,193 @@
+    
+    run(fout, date, yarn, verbose, inst, limit)
+
+    print('Start time  : %s' % time.strftime('%Y-%m-%d %H:%M:%S GMT', time.gmtime(time0)))
```
you can remove this now since `@info` decorator will do a job

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaigns.py:
```diff
@@ -0,0 +1,265 @@
+# CMSSpark modules
+from CMSSpark.spark_utils import dbs_tables, phedex_tables, print_rows
+from CMSSpark.spark_utils import spark_context, split_dataset
+from CMSSpark.utils import elapsed_time
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaigns.py:
```diff
@@ -0,0 +1,265 @@
+
+CAMPAIGNS_TIME_DATA_FILE = 'spark_exec_time_campaigns.txt'
+
+def get_options():
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaigns.py:
```diff
@@ -0,0 +1,265 @@
+
+    ctx.stop()
+
+def main():
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_campaigns.py:
```diff
@@ -0,0 +1,265 @@
+
+    run(fout, date, yarn, verbose, inst, limit)
+
+    print('Start time  : %s' % time.strftime('%Y-%m-%d %H:%M:%S GMT', time.gmtime(time0)))
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_dbs.py:
```diff
@@ -0,0 +1,128 @@
+# CMSSpark modules
+from CMSSpark.spark_utils import dbs_tables, phedex_tables, print_rows
+from CMSSpark.spark_utils import spark_context, split_dataset
+from CMSSpark.utils import elapsed_time
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_dbs.py:
```diff
@@ -0,0 +1,128 @@
+
+DBS_TIME_DATA_FILE = 'spark_exec_time_tier_dbs.txt'
+
+def get_options():
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_dbs.py:
```diff
@@ -0,0 +1,128 @@
+
+    ctx.stop()
+
+def main():
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_dbs.py:
```diff
@@ -0,0 +1,128 @@
+    
+    run(fout, yarn, verbose, inst)
+
+    print('Start time  : %s' % time.strftime('%Y-%m-%d %H:%M:%S GMT', time.gmtime(time0)))
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_leftovers.py:
```diff
@@ -0,0 +1,204 @@
+# CMSSpark modules
+from CMSSpark.spark_utils import dbs_tables, phedex_tables, print_rows
+from CMSSpark.spark_utils import spark_context, split_dataset
+from CMSSpark.utils import elapsed_time, bytes_to_readable
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_leftovers.py:
```diff
@@ -0,0 +1,204 @@
+
+LEFTOVERS_TIME_DATA_FILE = 'spark_exec_time_leftovers.txt'
+
+def get_options():
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_leftovers.py:
```diff
@@ -0,0 +1,204 @@
+
+    ctx.stop()
+
+def main():
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_leftovers.py:
```diff
@@ -0,0 +1,204 @@
+    
+    run(fout, date, yarn, verbose, inst)
+
+    print('Start time  : %s' % time.strftime('%Y-%m-%d %H:%M:%S GMT', time.gmtime(time0)))
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_phedex.py:
```diff
@@ -0,0 +1,135 @@
+# CMSSpark modules
+from CMSSpark.spark_utils import phedex_tables, print_rows
+from CMSSpark.spark_utils import spark_context, split_dataset
+from CMSSpark.utils import elapsed_time, split_date
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_phedex.py:
```diff
@@ -0,0 +1,135 @@
+
+PHEDEX_TIME_DATA_FILE = 'spark_exec_time_tier_phedex.txt'
+
+def get_options():
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_phedex.py:
```diff
@@ -0,0 +1,135 @@
+
+    ctx.stop()
+
+def main():
```
the same comment as in aggregate_campaign_tier.py

vkuznet commented on src/python/CMSSpark/reports/aggregate_phedex.py:
```diff
@@ -0,0 +1,135 @@
+
+    run(date, fout, yarn, verbose)
+
+    print('Start time  : %s' % time.strftime('%Y-%m-%d %H:%M:%S GMT', time.gmtime(time0)))
```
the same comment as in aggregate_campaign_tier.py

andrius-k replied:
Valentin, I didn't use options parser from conf.py because arguments list is different in each file. For example some files require --limit argument and others don't. And in conf.py file --no-log4j argument is defined which I never use. Do you want me to replace my option parsing code with the class defined in conf.py by adding --limit argument to conf.py and making proper use of --no-log4j in my scripts?

andrius-k replied:
Oh, I see that OptionsParser exposes ArgumentParser instance so I can add any additional arguments in each script, so that's good. But what to do with --no-log4j argument? I can see that in some of your scripts this argument is not used as well. Do you think that it's a good idea to have an argument in script that is not used?

vkuznet commented:
First, --no-log4j argument is not used in python directly but rather in
run_spark, so it can be used in all scripts to suppress the amount of loggin
from Java printouts.

Second, OptionParser contains main list of arguments used by all scripts,
all other can add additional arguments if necessary, see example here:
https://github.com/vkuznet/CMSSpark/blob/master/src/python/CMSSpark/dbs_cmssw.py#L91-95
which shows how to add --inst argument to script which requires it.
You can do the same with your limit argument and add it *only* in those
files which need it.

On  0, Andrius <notifications@github.com> wrote:
> andrius-k commented on this pull request.
> 
> 
> 
> > +
> +    ctx.stop()
> +
> +def main():
> +    "Main function"
> +    opts = get_options()
> +    print("Input arguments: %s" % opts)
> +    time0 = time.time()
> +    fout = opts.fout
> +    date = opts.date
> +    verbose = opts.verbose
> +    yarn = opts.yarn
> +
> +    run(date, fout, yarn, verbose)
> +
> +    print('Start time  : %s' % time.strftime('%Y-%m-%d %H:%M:%S GMT', time.gmtime(time0)))
> 
> Valentin, I didn't use options parser from conf.py because arguments list is different in each file. For example some files require --limit argument and others don't. And in conf.py file --no-log4j argument is defined which I never use. Do you want me to replace my option parsing code with the class defined in conf.py by adding --limit argument to conf.py and making proper use of --no-log4j in my scripts?
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/vkuznet/CMSSpark/pull/12#discussion_r176586405

andrius-k commented:
Oh, ok I can see that run_spark actually searches for no-log4j argument so it will be applicable to my scripts as well. Ok, I will refactor my code based on your comments.

andrius-k commented:
Valentin, I changed my code according to your request. I also added new `@info_save` decorator in `utils.py`. It takes file path as an argument and it saves elapsed time to that file. I need that time information for my reports.

andrius-k commented:
Valentin, I added usage info and examples on how to run each script. I also added `delete_trash` script to delete hdfs trashes.

vkuznet commented on src/bash/delete_trash:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
```
We don't need this if you'll use
hadoop fs -rm -r -skipTrash /hdfs_path command all the time.

andrius-k commented:
Valentin, I added `-skipTrash` argument to all scripts and removed `delete_trash` script entirely.

vkuznet commented on src/bash/report_campaigns/aggregate_campaign_tier:
```diff
@@ -0,0 +1,66 @@
+echo 'Results path: '$location
+
+# Results will be put here
+hdir=hdfs://$location
```
The location should be provided in a form of /path or hdfs:///path Please adjust code/usage to handle this case. If somehow user pass hdfs:///path it will fail here since you already defined a prefix.

andrius-k commented:
Valentin, ok, I will handle `hdfs://` prefix accordingly. Regarding the second part of you comment, are you saying that job will fail if provided path does not exist? I think the path will be created in such case but I will double check.

vkuznet commented:
For the second part, you can try it out I don't remember. But I think if you
submit a job and path already exists the job will fail. Since it depends on
script logic that's why I commented that you should try to accommodate your
logic.

On  0, Andrius <notifications@github.com> wrote:
> Valentin, ok, I will handle `hdfs://` prefix accordingly. Regarding the second part of you comment, are you saying that job will fail if provided path does not exist? I think the path will be created in such case but I will double check.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/vkuznet/CMSSpark/pull/12#issuecomment-376161289

andrius-k commented:
Yes, if path exists job fails, but that's exactly why I added `hadoop fs -rm -r -skipTrash $location` before I run every job. Do you think that this is not enough?

vkuznet commented:
Ahh, ok then.

On  0, Andrius <notifications@github.com> wrote:
> Yes, if path exists job fails, but that's exactly why I added `hadoop fs -rm -r -skipTrash $location` before I run every job. Do you think that this is not enough?
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/vkuznet/CMSSpark/pull/12#issuecomment-376174452

vkuznet commented:
so, is it ready now?

andrius-k commented:
Not yet, I am now working on fixes that Carl asked me to do. I can quickly fix `hdfs://` prefix issue but the code will still be half ready. I will inform you when I will be finished.

andrius-k commented:
Valentin, I fixed all issues reported by Carl and I fixed the `hdfs://` prefix issue.
